### ISSUE TITLE
Emit PossiblyInvalidPropertyAssignment instead of InvalidPropertyAssignment when types overlap?

### ISSUE THREAD

**Issue by TysonAndre** [vimeo/psalm#301]
Use case: There are a large number of false positives in a project due to the inferred types (correctly) including false/null. Some false positives include other types.

After splitting this up, InvalidPropertyAssignment would have a much lower false positive ratio, and issues could be fixed before PossiblyInvalidPropertyAssignment

Example: 

```php
<?php

class HasResource {
    /** @var resource */
    public $fout;

    public function __construct() {
        $this->fout = fopen('/tmp/psalm-output.txt', 'w');  // Observed: InvalidPropertyAssignment. Desired: PossiblyInvalidPropertyAssignment
        // Would also expect PossiblyInvalidPropertyAssignment when assigning string|false to property of type string, etc.
    }
}
$x = new HasResource();
```
